algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs

```diff
@@ -67,80 +64,148 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        alpha: &F,
+    ) -> DensePolynomial<F> {
+
+        let mut job_pool = ExecutionPool::with_capacity(state.circuit_specific_states.len());
+        let max_constraint_domain_size = state.max_constraint_domain.size_as_field_element;
+
+        for (i, (circuit, oracles)) in state.first_round_oracles.as_ref().unwrap().batches.iter().enumerate() {
+            let circuit_specific_state = &state.circuit_specific_states[circuit];
+            let summed_z_m = &mut summed_z_m_and_t[2*i];
+            assert!(*summed_z_m.0 == circuit.hash);
+            let summed_z_m = core::mem::take(&mut summed_z_m.1);
+            let summed_t = &mut summed_z_m_and_t[2*i + 1];
+            assert!(*summed_t.0 == circuit.hash);
+            let summed_t = core::mem::take(&mut summed_t.1);
+            let circuit_combiner = batch_combiners[&circuit.hash].circuit_combiner;
+            let instance_combiners = batch_combiners[&circuit.hash].instance_combiners.clone();
+            let x_polys = circuit_specific_state.x_polys.clone();
+            let input_domain = circuit_specific_state.input_domain.clone();
+            let constraint_domain = &circuit_specific_state.constraint_domain;
+            let fft_precomputation = &circuit.fft_precomputation;
+            let ifft_precomputation = &circuit.ifft_precomputation;
+            let circuit_id = &circuit.hash;
+
+            job_pool.add_job(move || {
+                let z_time = start_timer!(|| format!("Compute z poly for circuit {:?}", circuit_id));
+                let z = cfg_iter!(oracles)
+                    .zip_eq(instance_combiners)
+                    .zip(x_polys)
+                    .map(|((oracles, instance_combiner), x_poly)| {
+                        let mut z = oracles.w_poly.polynomial().as_dense().unwrap().mul_by_vanishing_poly(input_domain);
+                        // Zip safety: `x_poly` is smaller than `z_poly`.
+                        z.coeffs.iter_mut().zip(&x_poly.coeffs).for_each(|(z, x)| *z += x);
+                        cfg_iter_mut!(z.coeffs).for_each(|z| *z *= instance_combiner);
+                        z
+                    })
+                    .sum::<DensePolynomial<F>>();
+                assert!(z.degree() <= constraint_domain.size());
+
+                end_timer!(z_time);
+
+                let mut circuit_specific_lhs = Self::calculate_circuit_specific_lhs(constraint_domain, fft_precomputation, ifft_precomputation, summed_t, summed_z_m, z, *alpha);
+        
+                // Naive setup:
+                circuit_specific_lhs = circuit_specific_lhs.mul_by_vanishing_poly(state.max_constraint_domain);
+                let (quotient, remainder) = circuit_specific_lhs.divide_by_vanishing_poly(*constraint_domain).unwrap();
+                assert!(remainder.is_zero());
+                circuit_specific_lhs = quotient * (constraint_domain.size_as_field_element / max_constraint_domain_size);
```

Comment:
Let's compute `max_constraint_domain_size.inverse()` once and reuse it

Reply:
Good catch